### PR TITLE
Nulecule app support

### PR DESCRIPTION
**Pull request by sub-mod**
https://github.com/cockpit-project/cockpit/wiki/Feature:-Kubernetes:-Application-Entities

TESTING :

1)install atomic
git clone https://github.com/projectatomic/atomic-enterprise.git
python setup.py install
2)click on deploy button
3)select nulecule option
4)give image as submod/guestbookphp-app and give any namespace value
5)click deploy button in dialog

atomic run --opt2="--answers=/atomicapp/answers.conf"  submod/guestbookphp-app^C
atomic install --opt2="--answers-format=json" submod/guestbookphp-app

### CODE REVIEW

**Comment by stefwalter** [comment]
@sub-mod Does this replace #2332?

File: pkg/kubernetes/nulecule.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+        self.installrun = function installrun(ptype, tmp_dir, image) {
+            var deferred = $.Deferred();
+            var status = '';
+            var promise;
+            var buffer = '';
+            var process ;
+            
+            if(ptype === "install") {
+                process = cockpit.spawn(["atomicapp", "install", "--destination", tmp_dir, image],{ err: "out", superuser: true });
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Doesn't this need to be changed to use the 'atomic' command directly ... before we can merge this?

File: pkg/kubernetes/nulecule.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                debug("installing image: " + image + " in folder " + tmp_dir);
+            } else {
+                process = cockpit.spawn(['/bin/sh', '-s'],{ err: "out", superuser: true }).input("cd " + tmp_dir + " && /usr/bin/atomicapp -v run .");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Could you use "directory" option to spawn instead of 'cd' here?

File: pkg/kubernetes/nulecule.js (hunk 1 above)
Comment (sub-mod) [inline pkg/kubernetes/nulecule.js]: yes, i need to do that , atomic command is not working for me now.

**Comment by petervo** [comment]
I haven't looked too much into the underlying commands yet, so sorry if this way off or already covered in other discussions, but is there a way to not have to create a temporary dir? Being able to pipe to it somehow or parse the data and attach it as args. If that's not possible we need to make sure to clean up the tmp folder when we are done with it.

**Comment by sub-mod** [comment]
@petervo The tmp folder is quite important, since it holds quite a lot of files to deploy the application.
We can only delete it once the user doesn't need it.
I need to check where i can add the logic to delete the tmp folder.

**Comment by petervo** [comment]
Ok, i'd suggest them maybe encapsulating the directory creation within the install function storing it internally for the deploy step and deleting it close. That would also help simply the code in the client (deploy.js). WDYT?

**Comment by sub-mod** [comment]
@petervo , I added the code to delete the folder, tested it and it seems to work fine.

File: pkg/kubernetes/deploy.js [hunk 3]
```diff
@@ -77,6 +81,192 @@ define([
...
+                deferred.reject(exs);
+                $("#deploy-app-dialog").dialog("failure", exs);
+            })
+            .done(function(fields) {
+                nulecule_client.install(fields.nulecule_image)
+                    .progress(function(msg) {
+                        deferred.notify(msg);
+                    })
+                    .done(function(){
+                        var answerfile = nulecule_client.answers;
```
Comment (petervo) [inline pkg/kubernetes/deploy.js]: WDYT, of returning this as a result and then passing it back in the run method. Or is that too complicated?

File: pkg/kubernetes/deploy.js [hunk 4]
```diff
@@ -77,6 +81,192 @@ define([
...
+            ex = new Error(_("Please provide a valid namespace."));
+        if (ex) {
+            ex.target = "#deploy-app-namespace-group";
+            fails.push(ex);
+            ex = null;
+        } else {
+            answer_data.general.namespace = ns;
+        }
+
+        var nulecule_image = $("#deploy-app-nulecule-image");
```
Comment (petervo) [inline pkg/kubernetes/deploy.js]: Shouldn't we require that this stays the same between run and install? Or does it make sense for users to be allowed to change this.

File: pkg/kubernetes/nulecule.js [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+            debug("removing folder " + folder);
+            var process = cockpit.spawn(["rm", "-fr", folder],{ err: "out", superuser: true , directory: '/tmp', pty: true });
+            install_dir = null;
+
+            return process.fail(function (ex) {
+                console.warn("Error cleaning up nulecule temporary files: " + ex.message);
+            });
+        }
+
+	   function write_answer_file(data) {
```
Comment (petervo) [inline pkg/kubernetes/nulecule.js]: indentation

File: pkg/kubernetes/nulecule.js [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                    var msgl;
+                    if(text.indexOf("atomicapp.status.info.message") > -1) {
+                        msgl = text.split('atomicapp.status.info.message=');
+                        if(msgl.length > 1)
+                            deferred.notify(msgl[1]);
+
+                    } else if(text.indexOf("atomicapp.status.answer.message") > -1) {
+                        msgl = text.split('atomicapp.status.answer.message=');
+                        if(msgl.length > 1) {
+                            self.answers = JSON.parse(msgl[1]);
```
Comment (petervo) [inline pkg/kubernetes/nulecule.js]: Got some JSON parse errors here when running this. Do we have any way of fixing the server side code so we get valid json?

**Comment by petervo** [comment]
I think we should probably wait to merge this until the atomic changes are available on at least rawhide. Would also be nice to have an integration test.

File: pkg/kubernetes/deploy.js (hunk 4 above)
Comment (sub-mod) [inline pkg/kubernetes/deploy.js]: No, they should not be changed.
I have added code to disable them after install step.

**Comment by petervo** [comment]
I still can't get this working, is submod/guestbookphp-app still the right app to try?

**Comment by sub-mod** [comment]
@petervo i forgot to push the new docker images.can you try now please.

**Comment by petervo** [comment]
I was finally able to run this successfully after downgrading to docker 1.6 and building and tweeking the containers in #2521.

This still probably shouldn't go in until the the all the requirements make it to their upstreams so latest versions of docker, atomic and the standard sample nulecule apps work.

**Comment by stefwalter** [comment]
@mvollmer Can we get the 'atomic' package from updates-testing into our F22 and F23 test images? That would allow us to merge this.

**Comment by mvollmer** [comment]
> @mvollmer Can we get the 'atomic' package from updates-testing into our F22 and F23 test images? That would allow us to merge this.

Could we wait until it is in 'updates'?  It's almost there, just stuck in bureaucracy, I think.

**Comment by mvollmer** [comment]
@stefwalter, there doesn't seem to be any integration here that would actually need "atomic" in the test images.  So I don't think we need to do anything urgently, right?

**Comment by stefwalter** [comment]
Filed #2557 #2558 and #2559 as follow ups for this pull request. As discussed on IRC, we can merge this.

**Pull request closed without merging**